hermes/probe/monitor.go

```diff
@@ -12,9 +12,57 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 //
+// Authors:
+// - Alicja Kwiecinska (kwiecinskaa@google.com) github: alicjakwie
+// - Evan Spendlove, GitHub: evanSpendlove.
+//
 // Monitor implements a probe that monitors a storage system using the
 // Hermes algorithm.
 
+package probe
+
+import (
+	"fmt"
+
+	"github.com/google/cloudprober/logger"
+	"github.com/google/cloudprober/metrics"
+	"github.com/google/cloudprober/probes/options"
+
+	pb "github.com/googleinterns/step224-2020/config/proto"
+)
+
 // MonitorProbe holds aggregate information about all probe runs, per-target.
 // It also holds the config and options used to initialise the probe.
-type MonitorProbe struct{}
+type MonitorProbe struct {
+	name    string
+	config  *pb.HermesProbeDef
+	targets []*pb.Target
+	opts    *options.Options
+
+	// probeOpLatency is used to record latency values with distinct labels
+	// per probe operation per target.
+	// Recommended usage: append(probeOpLatency[target][probeOp], newMetric)
+	probeOpLatency map[string]map[string][]*metrics.EventMetrics
+	// apiCallLatency is used to record latency values with distinct labels
+	// per API call per target.
+	// Recommended usage: append(apiCallLatency[target][apiCall], newMetric)
+	apiCallLatency map[string]map[string][]*metric.EventMetrics
+	logger         *logger.Logger
+}
+
+// Init initializes the probe with the given parameters.
+// This is a required method to implement the cloudprober.Probes.Probe interface.
+func (mp *MonitorProbe) Init(name string, opts *options.Options) error {
+	conf, ok := opts.ProbeConf.(*pb.HermesProbeDef) // Casting ProbeConf interface to HermesProbeDef type
+	if !ok {
+		return fmt.Errorf("invalid argument: opts.ProbeConf is not of type *pb.HermesProbeDef")
+	}
+	mp.config = conf
+	mp.name = name
+	mp.opts = opts
+	mp.probeOpLatency = make(map[string]map[string]*metrics.EventMetrics)
+	mp.apiCallLatency = make(map[string]map[string]*metrics.EventMetrics)
```

Comment:
If you know the list of metric labels you'll have, then it's considered good practise in prometheus to initialise them to empty metrics.

Reply:
I'm not quite sure what you mean by this. Do you mean to add the labels without values for each metric?

Metric labels are initialised in the `initProbeMetrics()` function.

Reply:
I think I understand what you mean, and this is done when initialising the metrics.